File: Go-Solutions/1.go
package main

func twoSum1(nums []int, target int) []int {
	helperMap := make(map[int]int)
	for i, v := range nums {
		complement := target - v
		if j, ok := helperMap[complement]; ok && (j != i) {
			return []int{i, j}
		}
		helperMap[v] = i
	}
	panic("No solution")
}

File: Java-Kotlin/src/main/java/array/_1.java
package array;

import java.util.HashMap;
import java.util.Map;

/**
 * Problem link: https://leetcode.com/problems/two-sum/
 * <p>
 * SolutionDFS idea: Maintain a map value -> index
 * <p>
 * Time complexity: O(n)
 * <p>
 * Space complexity: O(n)
 */
public class _1 {

    public int[] twoSum(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            throw new IllegalArgumentException("invalid input");
        }
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement) && (map.get(complement) != i)) {
                return new int[]{i, map.get(complement)};
            }
            map.put(nums[i], i);
        }
        throw new IllegalArgumentException("no solution ");
    }
}
